forest/middlewares.py

```diff
@@ -0,0 +1,15 @@
+"""Middleware utilities"""
+
+
+class Log:
+    def __init__(self):
+        self.actions = []
+
+    def __call__(self, store, action):
+        self.actions.append(action)
+        yield action
```

Comment:
It'd be nice if we could avoid needing this...

Reply:
In particular, I'm referring to the tests. It'd be nice to use a pattern more like the one you suggested earlier.

Reply:
Agreed, this was a convenience data structure under the old implementation but now it is just technical debt and considered an anti-pattern/memory leak in the new implementation